test failure due to data race

```
==================
WARNING: DATA RACE
Read at 0x00c4201f8218 by goroutine 36:
  github.com/runconduit/conduit/controller/api/public.(*mockGrpcServer).Tap()
      /home/travis/gopath/src/github.com/runconduit/conduit/controller/api/public/http_server_test.go:53 +0x115
  github.com/runconduit/conduit/controller/api/public.(*handler).handleTap()
      /home/travis/gopath/src/github.com/runconduit/conduit/controller/api/public/http_server.go:158 +0x18b
  github.com/runconduit/conduit/controller/api/public.(*handler).ServeHTTP()
      /home/travis/gopath/src/github.com/runconduit/conduit/controller/api/public/http_server.go:50 +0x4db
  net/http.serverHandler.ServeHTTP()
      /home/travis/.gimme/versions/go1.9.linux.amd64/src/net/http/server.go:2619 +0xbc
  net/http.(*conn).serve()
      /home/travis/.gimme/versions/go1.9.linux.amd64/src/net/http/server.go:1801 +0x83b
Previous write at 0x00c4201f8218 by goroutine 48:
  github.com/runconduit/conduit/controller/api/public.TestServer.func4()
      /home/travis/gopath/src/github.com/runconduit/conduit/controller/api/public/http_server_test.go:179 +0xcd
  testing.tRunner()
      /home/travis/.gimme/versions/go1.9.linux.amd64/src/testing/testing.go:746 +0x16c
Goroutine 36 (running) created at:
  net/http.(*Server).Serve()
      /home/travis/.gimme/versions/go1.9.linux.amd64/src/net/http/server.go:2720 +0x37c
  net/http.Serve()
      /home/travis/.gimme/versions/go1.9.linux.amd64/src/net/http/server.go:2323 +0xe2
  github.com/runconduit/conduit/controller/api/public.TestServer.func1()
      /home/travis/gopath/src/github.com/runconduit/conduit/controller/api/public/http_server_test.go:75 +0x5c
Goroutine 48 (finished) created at:
  testing.(*T).Run()
      /home/travis/.gimme/versions/go1.9.linux.amd64/src/testing/testing.go:789 +0x568
  github.com/runconduit/conduit/controller/api/public.TestServer()
      /home/travis/gopath/src/github.com/runconduit/conduit/controller/api/public/http_server_test.go:178 +0x58a
  testing.tRunner()
      /home/travis/.gimme/versions/go1.9.linux.amd64/src/testing/testing.go:746 +0x16c
==================
==================
WARNING: DATA RACE
Read at 0x00c420466de0 by goroutine 36:
  errors.(*errorString).Error()
      /home/travis/.gimme/versions/go1.9.linux.amd64/src/errors/errors.go:19 +0x3b
  github.com/runconduit/conduit/controller/api/public.writeErrorToHttpResponse()
      /home/travis/gopath/src/github.com/runconduit/conduit/controller/api/public/proto_over_http.go:71 +0x163
  github.com/runconduit/conduit/controller/api/public.(*handler).handleTap()
      /home/travis/gopath/src/github.com/runconduit/conduit/controller/api/public/http_server.go:160 +0x1c2
  github.com/runconduit/conduit/controller/api/public.(*handler).ServeHTTP()
      /home/travis/gopath/src/github.com/runconduit/conduit/controller/api/public/http_server.go:50 +0x4db
  net/http.serverHandler.ServeHTTP()
      /home/travis/.gimme/versions/go1.9.linux.amd64/src/net/http/server.go:2619 +0xbc
  net/http.(*conn).serve()
      /home/travis/.gimme/versions/go1.9.linux.amd64/src/net/http/server.go:1801 +0x83b
Previous write at 0x00c420466de0 by goroutine 48:
  github.com/runconduit/conduit/controller/api/public.TestServer.func4()
      /home/travis/gopath/src/github.com/runconduit/conduit/controller/api/public/client.go:5 +0x93
  testing.tRunner()
      /home/travis/.gimme/versions/go1.9.linux.amd64/src/testing/testing.go:746 +0x16c
Goroutine 36 (running) created at:
  net/http.(*Server).Serve()
      /home/travis/.gimme/versions/go1.9.linux.amd64/src/net/http/server.go:2720 +0x37c
  net/http.Serve()
      /home/travis/.gimme/versions/go1.9.linux.amd64/src/net/http/server.go:2323 +0xe2
  github.com/runconduit/conduit/controller/api/public.TestServer.func1()
      /home/travis/gopath/src/github.com/runconduit/conduit/controller/api/public/http_server_test.go:75 +0x5c
Goroutine 48 (finished) created at:
  testing.(*T).Run()
      /home/travis/.gimme/versions/go1.9.linux.amd64/src/testing/testing.go:789 +0x568
  github.com/runconduit/conduit/controller/api/public.TestServer()
      /home/travis/gopath/src/github.com/runconduit/conduit/controller/api/public/http_server_test.go:178 +0x58a
  testing.tRunner()
      /home/travis/.gimme/versions/go1.9.linux.amd64/src/testing/testing.go:746 +0x16c
==================
=== RUN   TestDeserializePayloadFromReader/Returns_error_when_message_has_fewer_bytes_than_declared_message_size
--- FAIL: TestDeserializePayloadFromReader (0.05s)
    --- PASS: TestDeserializePayloadFromReader/Can_read_message_correctly_based_on_payload_size_correct_payload_size_to_message (0.00s)
    --- PASS: TestDeserializePayloadFromReader/Can_multiple_messages_in_the_same_stream (0.03s)
    --- PASS: TestDeserializePayloadFromReader/Can_write_and_read__marshalled_protobuf_messages (0.01s)
    --- PASS: TestDeserializePayloadFromReader/Can_read_byte_streams_larger_than_Go's_default_buffer_chunk_size (0.02s)
    --- PASS: TestDeserializePayloadFromReader/Returns_error_when_message_has_fewer_bytes_than_declared_message_size (0.00s)
	testing.go:699: race detected during execution of test
```

--- a/cli/cmd/install.go
+++ b/cli/cmd/install.go
@@ -42,6 +42,7 @@ type (
 		highAvailability            bool
 		controllerUID               int64
 		disableH2Upgrade            bool
+		disableHeartbeat            bool
 		noInitContainer             bool
 		skipChecks                  bool
 		omitWebhookSideEffects      bool
@@ -163,6 +164,7 @@ func newInstallOptionsWithDefaults() (*installOptions, error) {
 		highAvailability:            defaults.HighAvailability,
 		controllerUID:               defaults.ControllerUID,
 		disableH2Upgrade:            !defaults.EnableH2Upgrade,
+		disableHeartbeat:            defaults.DisableHeartBeat,
 		noInitContainer:             defaults.NoInitContainer,
 		omitWebhookSideEffects:      defaults.OmitWebhookSideEffects,
 		restrictDashboardPrivileges: defaults.RestrictDashboardPrivileges,
@@ -430,6 +432,10 @@ func (options *installOptions) recordableFlagSet() *pflag.FlagSet {
 		&options.disableH2Upgrade, "disable-h2-upgrade", options.disableH2Upgrade,
 		"Prevents the controller from instructing proxies to perform transparent HTTP/2 upgrading (default false)",
 	)
+	flags.BoolVar(
+		&options.disableHeartbeat, "disable-heartbeat", options.disableHeartbeat,
+		"Disables the heartbeat cronjob (default false)",
+	)
 	flags.DurationVar(
 		&options.identityOptions.issuanceLifetime, "identity-issuance-lifetime", options.identityOptions.issuanceLifetime,
 		"The amount of time for which the Identity issuer should certify identity",
@@ -626,6 +632,7 @@ func (options *installOptions) buildValuesWithoutIdentity(configs *pb.All) (*cha
 	installValues.PrometheusLogLevel = toPromLogLevel(strings.ToLower(options.controllerLogLevel))
 	installValues.HeartbeatSchedule = options.heartbeatSchedule()
 	installValues.RestrictDashboardPrivileges = options.restrictDashboardPrivileges
+	installValues.DisableHeartBeat = options.disableHeartbeat
 	installValues.UUID = configs.GetInstall().GetUuid()
 	installValues.WebImage = fmt.Sprintf("%s/web", options.dockerRegistry)
 

--- a/cli/cmd/upgrade.go
+++ b/cli/cmd/upgrade.go
@@ -70,7 +70,7 @@ func newCmdUpgradeConfig(options *upgradeOptions) *cobra.Command {
 
 Note that this command should be followed by "linkerd upgrade control-plane".`,
 		Example: `  # Default upgrade.
-  linkerd upgrade config | kubectl apply -f -`,
+  linkerd upgrade config | kubectl apply --prune -l linkerd.io/control-plane-ns=linkerd -f -`,
 		RunE: func(cmd *cobra.Command, args []string) error {
 			return upgradeRunE(options, configStage, options.recordableFlagSet())
 		},
@@ -95,7 +95,7 @@ Note that the default flag values for this command come from the Linkerd control
 plane. The default values displayed in the Flags section below only apply to the
 install command. It should be run after "linkerd upgrade config".`,
 		Example: `  # Default upgrade.
-  linkerd upgrade control-plane | kubectl apply -f -`,
+  linkerd upgrade control-plane | kubectl apply --prune -l linkerd.io/control-plane-ns=linkerd -f -`,
 		RunE: func(cmd *cobra.Command, args []string) error {
 			return upgradeRunE(options, controlPlaneStage, flags)
 		},
@@ -126,7 +126,7 @@ plane. The default values displayed in the Flags section below only apply to the
 install command.`,
 
 		Example: `  # Default upgrade.
-  linkerd upgrade | kubectl apply -f -
+  linkerd upgrade | kubectl apply --prune -l linkerd.io/control-plane-ns=linkerd -f -
 
   # Similar to install, upgrade may also be broken up into two stages, by user
   # privilege.`,

--- a/pkg/healthcheck/healthcheck.go
+++ b/pkg/healthcheck/healthcheck.go
@@ -539,7 +539,7 @@ func (hc *HealthChecker) allCategories() []category {
 					hintAnchor:  "l5d-existence-sa",
 					fatal:       true,
 					check: func(context.Context) error {
-						return hc.checkServiceAccounts(true)
+						return hc.checkServiceAccounts(false)
 					},
 				},
 				{
@@ -586,6 +586,14 @@ func (hc *HealthChecker) allCategories() []category {
 					check: func(context.Context) (err error) {
 						hc.linkerdConfig, err = hc.checkLinkerdConfigConfigMap()
 						return
+					},
+				},
+				{
+					description: "control plane ServiceAccounts exist",
+					hintAnchor:  "l5d-existence-sa",
+					fatal:       true,
+					check: func(context.Context) error {
+						return hc.checkServiceAccounts(true)
 					},
 				},
 				{
@@ -1033,11 +1041,13 @@ func (hc *HealthChecker) expectedRBACNames() []string {
 	}
 }
 
-func expectedServiceAccountNames() []string {
+func expectedServiceAccountNames(onlyHeartbeat bool) []string {
+	if onlyHeartbeat {
+		return []string{"linkerd-heartbeat"}
+	}
 	return []string{
 		"linkerd-controller",
 		"linkerd-grafana",
-		"linkerd-heartbeat",
 		"linkerd-identity",
 		"linkerd-prometheus",
 		"linkerd-proxy-injector",
@@ -1093,7 +1103,7 @@ func (hc *HealthChecker) checkClusterRoleBindings(shouldExist bool) error {
 	return checkResources("ClusterRoleBindings", objects, hc.expectedRBACNames(), shouldExist)
 }
 
-func (hc *HealthChecker) checkServiceAccounts(shouldExist bool) error {
+func (hc *HealthChecker) checkServiceAccounts(onlyHeartbeat bool) error {
 	options := metav1.ListOptions{
 		LabelSelector: k8s.ControllerNSLabel,
 	}
@@ -1108,7 +1118,20 @@ func (hc *HealthChecker) checkServiceAccounts(shouldExist bool) error {
 		objects = append(objects, &item)
 	}
 
-	return checkResources("ServiceAccounts", objects, expectedServiceAccountNames(), shouldExist)
+	if onlyHeartbeat {
+		heartbeatDisabled := false
+		for _, flag := range hc.linkerdConfig.GetInstall().GetFlags() {
+			if flag.GetName() == "disable-heartbeat" && flag.GetValue() == "true" {
+				heartbeatDisabled = true
+				break
+			}
+		}
+		if heartbeatDisabled {
+			return nil
+		}
+	}
+
+	return checkResources("ServiceAccounts", objects, expectedServiceAccountNames(onlyHeartbeat), true)
 }
 
 func (hc *HealthChecker) checkCustomResourceDefinitions(shouldExist bool) error {

--- a/pkg/healthcheck/healthcheck.go
+++ b/pkg/healthcheck/healthcheck.go
@@ -134,6 +134,17 @@ const AllowedClockSkew = time.Minute + tls.DefaultClockSkewAllowance
 var (
 	retryWindow    = 5 * time.Second
 	requestTimeout = 30 * time.Second
+
+	expectedServiceAccountNames = []string{
+		"linkerd-controller",
+		"linkerd-grafana",
+		"linkerd-identity",
+		"linkerd-prometheus",
+		"linkerd-proxy-injector",
+		"linkerd-sp-validator",
+		"linkerd-web",
+		"linkerd-tap",
+	}
 )
 
 // Resource provides a way to describe a Kubernetes object, kind, and name.
@@ -539,7 +550,7 @@ func (hc *HealthChecker) allCategories() []category {
 					hintAnchor:  "l5d-existence-sa",
 					fatal:       true,
 					check: func(context.Context) error {
-						return hc.checkServiceAccounts(false)
+						return hc.checkServiceAccounts(expectedServiceAccountNames)
 					},
 				},
 				{
@@ -589,11 +600,14 @@ func (hc *HealthChecker) allCategories() []category {
 					},
 				},
 				{
-					description: "control plane ServiceAccounts exist",
+					description: "heartbeat ServiceAccount exist",
 					hintAnchor:  "l5d-existence-sa",
 					fatal:       true,
 					check: func(context.Context) error {
-						return hc.checkServiceAccounts(true)
+						if hc.isHeartbeatDisabled() {
+							return nil
+						}
+						return hc.checkServiceAccounts([]string{"linkerd-heartbeat"})
 					},
 				},
 				{
@@ -1041,22 +1055,6 @@ func (hc *HealthChecker) expectedRBACNames() []string {
 	}
 }
 
-func expectedServiceAccountNames(onlyHeartbeat bool) []string {
-	if onlyHeartbeat {
-		return []string{"linkerd-heartbeat"}
-	}
-	return []string{
-		"linkerd-controller",
-		"linkerd-grafana",
-		"linkerd-identity",
-		"linkerd-prometheus",
-		"linkerd-proxy-injector",
-		"linkerd-sp-validator",
-		"linkerd-web",
-		"linkerd-tap",
-	}
-}
-
 func (hc *HealthChecker) checkClusterRoles(shouldExist bool) error {
 	options := metav1.ListOptions{
 		LabelSelector: k8s.ControllerNSLabel,
@@ -1103,7 +1101,16 @@ func (hc *HealthChecker) checkClusterRoleBindings(shouldExist bool) error {
 	return checkResources("ClusterRoleBindings", objects, hc.expectedRBACNames(), shouldExist)
 }
 
-func (hc *HealthChecker) checkServiceAccounts(onlyHeartbeat bool) error {
+func (hc *HealthChecker) isHeartbeatDisabled() bool {
+	for _, flag := range hc.linkerdConfig.GetInstall().GetFlags() {
+		if flag.GetName() == "disable-heartbeat" && flag.GetValue() == "true" {
+			return true
+		}
+	}
+	return false
+}
+
+func (hc *HealthChecker) checkServiceAccounts(saNames []string) error {
 	options := metav1.ListOptions{
 		LabelSelector: k8s.ControllerNSLabel,
 	}
@@ -1118,20 +1125,7 @@ func (hc *HealthChecker) checkServiceAccounts(onlyHeartbeat bool) error {
 		objects = append(objects, &item)
 	}
 
-	if onlyHeartbeat {
-		heartbeatDisabled := false
-		for _, flag := range hc.linkerdConfig.GetInstall().GetFlags() {
-			if flag.GetName() == "disable-heartbeat" && flag.GetValue() == "true" {
-				heartbeatDisabled = true
-				break
-			}
-		}
-		if heartbeatDisabled {
-			return nil
-		}
-	}
-
-	return checkResources("ServiceAccounts", objects, expectedServiceAccountNames(onlyHeartbeat), true)
+	return checkResources("ServiceAccounts", objects, saNames, true)
 }
 
 func (hc *HealthChecker) checkCustomResourceDefinitions(shouldExist bool) error {

--- a/test/testdata/check.golden
+++ b/test/testdata/check.golden
@@ -22,7 +22,7 @@ linkerd-config
 linkerd-existence
 -----------------
 √ 'linkerd-config' config map exists
-√ control plane ServiceAccounts exist
+√ heartbeat ServiceAccount exist
 √ control plane replica sets are ready
 √ no unschedulable pods
 √ controller pod is running

--- a/test/testdata/check.golden
+++ b/test/testdata/check.golden
@@ -22,6 +22,7 @@ linkerd-config
 linkerd-existence
 -----------------
 √ 'linkerd-config' config map exists
+√ control plane ServiceAccounts exist
 √ control plane replica sets are ready
 √ no unschedulable pods
 √ controller pod is running

--- a/test/testdata/check.proxy.golden
+++ b/test/testdata/check.proxy.golden
@@ -22,7 +22,7 @@ linkerd-config
 linkerd-existence
 -----------------
 √ 'linkerd-config' config map exists
-√ control plane ServiceAccounts exist
+√ heartbeat ServiceAccount exist
 √ control plane replica sets are ready
 √ no unschedulable pods
 √ controller pod is running

--- a/test/testdata/check.proxy.golden
+++ b/test/testdata/check.proxy.golden
@@ -22,6 +22,7 @@ linkerd-config
 linkerd-existence
 -----------------
 √ 'linkerd-config' config map exists
+√ control plane ServiceAccounts exist
 √ control plane replica sets are ready
 √ no unschedulable pods
 √ controller pod is running

